Describe the comments module (#2278)
reimagined/resolve: docs/modules/comments.md

---
id: comments
title: Comments
---

# Comments Module

The reSolve comments module ([@resolve-js/module-comments](https://www.npmjs.com/package/@resolve-js/module-comments)) adds support for hierarchical user comments in a reSolve application. In addition to aggregates and read models that describe the comments logic on the server side, the module exports renderless client components based on React + Redux and the [@resolve-js/redux](https://www.npmjs.com/package/@resolve-js/@resolve-js/redux) library that you can use to allow your application's users to post comments and to display these comments on a page.

## Installation

Use the following console input to install the uploader module:

```sh
yarn add @resolve-js/module-comments
```

## Register and Configure the Module

Register the installed module in the project's `run.js` file:

```js
import { merge } from '@resolve-js/scripts'
import createModuleComments from '@resolve-js/module-comments'

// All of the options specified below are optional.

> const moduleComments = resolveModuleComments({

Reviewer: It looks like it should be the createModuleComments call instead of resolveModuleComments, see the import expression above

Author: Fixed

### `eventTypes`

The `eventTypes` option is an object whose fields specify custom names for the comments module's events. This object can contain all or some of the following fields:

```js
{
  COMMENT_CREATED,
  COMMENT_UPDATED,
  COMMENT_REMOVED,
}
```

### `commandTypes`

The `commandTypes` option is an object whose fields specify custom names for the comments module's commands. This object can contain all or some of the following fields:

```js
{
  createComment,
  updateComment,
  removeComment,
}
```

### `resolverNames`

The `resolverNames` option is an object whose fields specify custom names for the comments read model's resolvers. This object can contain all or some of the following fields:

```js
  commentsTree,          // Returns a hierarchical tree of comments.
  foreignCommentsCount,  // Returns the number of comments left by other users.
  allCommentsPaginate,   // Returns comments with pagination.
```

### `verifyCommand`

The `verifyCommand` option accepts a function of the following signature:

```js
verifyCommand(state, command, jwt) {

>   // Place your custom ligic here.

Reviewer: Typo: 'ligic' instead of 'logic'

Author: Fixed

>   // Throw and error if verification fails.

Reviewer: Typo: 'and error' instead of 'an error'

Author: Fixed

```js
{
  authorId,  // The comment author's unique ID.
  commentId, // The new comment's unique ID.
}
```

### Resolvers

The comments module's read model (named `"Comments"` by default) has the following resolvers:

| Default Name                                    | Description                                         |
| ----------------------------------------------- | --------------------------------------------------- |
| [`commentsTree`](#commentstree)                 | Returns a hierarchical tree of comments.            |
| [`foreignCommentsCount`](#foreigncommentscount) | Returns the number of comments left by other users. |
| [`allCommentsPaginate`](#allcommentspaginate)   | Returns comments with pagination.                   |

#### `commentsTree`

Returns a hierarchical tree of comments.

**Arguments**

| Argument Name     | Description                                                     |
| ----------------- | --------------------------------------------------------------- |
| `treeId`          | The unique identifier of a comment tree.                        |
| `parentCommentId` | The identifier of a comment for which to obtain child comments. |

> | `authorId`        | The identifier of user whose comments to obtain.                |

Reviewer: 'of a user'

Author: Fixed

| ----------------- | -------------------------------------------------------------------- |
| `checkInterval`   | The time interval in milliseconds between checks for updates.        |
| `readModelName`   | The name of the comments read model.                                 |
| `resolverName`    | The name of the comments read model's resolver.                      |
| `onClick`         | A callback function that handles a user's click on the notification. |
| `treeId`          | The unique identifier of the current comment tree.                   |
| `parentCommentId` | The unique identifier of the parent comment.                         |
| `authorId`        | The comment author's unique identifier.                              |

This component passes all its props to its direct child.

The child component also receives the following additional props:

| Prop Name | Description                                                                                                               |
| --------- | ------------------------------------------------------------------------------------------------------------------------- |

> | `onClick` | Resets the notification and calls the `CommentsNotificationRenderless` component's `onClink` callback if it is specified. |

Reviewer: Typo: 'onClink' instead of 'onClick'

Author: Fixed

---
id: comments
title: Comments
---

```mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```

# Comments Module

The reSolve comments module ([@resolve-js/module-comments](https://www.npmjs.com/package/@resolve-js/module-comments)) adds support for user comments to a reSolve application. It implements an aggregate and a read model that together describe the logic used to process hierarchical trees of comments and serve them to a client.

> The comments module also exports renderless [client components](#client-api) based on React + Redux and the [@resolve-js/redux](https://www.npmjs.com/package/@resolve-js/@resolve-js/redux) library that you can use to implement a user interface required to post and navigate comments on your web application's page.

Reviewer 2: ```suggestion
The comments module also exports renderless [client components](#client-api) based on React+Redux and the [@resolve-js/redux](https://www.npmjs.com/package/@resolve-js/@resolve-js/redux) library that you can use to implement a user interface required to post and navigate comments on your web application's page.
```

  foreignCommentsCount,  // Returns the number of comments left by other users.
  allCommentsPaginate,   // Returns comments with pagination.
```

### `verifyCommand`

The `verifyCommand` option accepts a callback function of the following signature:

<!-- prettier-ignore-start -->

```js
(state, command, jwt) => {
  // Place your custom logic here.
  // Throw an error if verification fails.
}
```

<!-- prettier-ignore-end -->

## Server API

### Commands

The following command types are handled by the comments module's aggregate:

| Default Name                      | Description                  |
| --------------------------------- | ---------------------------- |
| [`createComment`](#createcomment) | Creates a new comment.       |
| [`updateComment`](#updatecomment) | Updates an existing comment. |
| [`removeComment`](#removecomment) | Removes an existing comment. |

#### `createComment`

Creates a new comment.

**Payload**

The `createComment` command requires a payload object of the following structure:

```js
{
  authorId,        // The comment author's unique ID.
  commentId,       // The new comment's unique ID.

>   parentCommentId, // (optional) The ID of the parrent comment.

Reviewer 2: ```suggestion
  parentCommentId, // Optional. The ID of the parrent comment.
```

```js
{
  authorId,  // The comment author's unique ID.
  commentId, // The comment's unique ID.
}
```

### Resolvers

The comments module's read model (named `"Comments"` by default) has the following resolvers:

| Default Name                                    | Description                                         |
| ----------------------------------------------- | --------------------------------------------------- |
| [`commentsTree`](#commentstree)                 | Returns a hierarchical tree of comments.            |
| [`foreignCommentsCount`](#foreigncommentscount) | Returns the number of comments left by other users. |
| [`allCommentsPaginate`](#allcommentspaginate)   | Returns comments with pagination.                   |

#### `commentsTree`

Returns a hierarchical tree of comments.

**Arguments**

| Argument Name     | Description                                                     |
| ----------------- | --------------------------------------------------------------- |
| `treeId`          | The unique identifier of a comment tree.                        |
| `parentCommentId` | The identifier of a comment for which to obtain child comments. |

> | `authorId`        | The identifier of a user whose comments to obtain.              |

Reviewer 2: ```suggestion
| `authorId`        | The identifier of a user whose comments are obtained.              |
```

| `maxLevel`        | The maximum nesting level of the comments to load.              |

#### `foreignCommentsCount`

Returns the number of comments left by other users.

**Arguments**

| Argument Name     | Description                                                     |
| ----------------- | --------------------------------------------------------------- |
| `treeId`          | The unique identifier of a comment tree.                        |
| `parentCommentId` | The identifier of a comment for which to obtain child comments. |
| `maxLevel`        | The maximum nesting level of the comments to load.              |

#### `allCommentsPaginate`

Returns comments with pagination.

**Arguments**

| Argument Name | Description                       |
| ------------- | --------------------------------- |
| `itemsOnPage` | The number of comments on a page. |
| `pageNumber`  | The number of the page to load.   |

## Client API

> The comments ([@resolve-js/module-comments](https://www.npmjs.com/package/@resolve-js/module-comments)) package exports renderless React+Redux components that you can use implement your own UI for user comments. The following components are available:

Reviewer 2: ```suggestion
The comments ([@resolve-js/module-comments](https://www.npmjs.com/package/@resolve-js/module-comments)) package exports renderless React+Redux components that you can use to implement your own UI for user comments. The following components are available:
```

| [`CommentsPaginateRenderless`](#commentspaginaterenderless)         | A container used to render comments in a paginated view.                         |
| [`CommentsTreeRenderless`](#resolvernames)                          | A container used to render comments in a hierarchical view.                      |

### `CommentsNotificationRenderless`

A component that displays a notification about new comments.

**Usage**

```js
import React from 'react'
import { CommentsNotificationRenderless } from '@resolve-js/module-comments'

export const CommentsNotification = (props) => (
  <CommentsNotificationRenderless
    treeId="treeId"
    parentCommentId="parentCommentId"
    authorId="authorId"
    checkInterval={10000} // default = 30 * 1000
    readModelName="customReadModelName" // default = 'readModelName'
    resolverName="customResolverName" // default = 'foreignCommentsCount'
    {...props}
  >
    {({ count, onClick }) => {
      // Replace with your code.
      if (count === 0) return null
      return (
        <div onClick={onClick}>

>           Comments have been updated, refresh the page to see the new comments.

Reviewer 2: Is this message editable? If so, can you split it into two sentences?

`Comments have been updated. Refresh the page to see the new comments.`

Author: I updated the message.

Reviewer 2: Thank you.

| `checkInterval`   | The time interval in milliseconds between checks for updates.        |
| `readModelName`   | The name of the comments read model.                                 |
| `resolverName`    | The name of the comments read model's resolver.                      |
| `onClick`         | A callback function that handles a user's click on the notification. |
| `treeId`          | The unique identifier of the current comment tree.                   |
| `parentCommentId` | The unique identifier of the parent comment.                         |
| `authorId`        | The comment author's unique identifier.                              |

This component passes all its props to its direct child.

The child component also receives the following additional props:

| Prop Name | Description                                                                                                               |
| --------- | ------------------------------------------------------------------------------------------------------------------------- |
| `onClick` | Resets the notification and calls the `CommentsNotificationRenderless` component's `onClick` callback if it is specified. |

> | `count`   | The number of available new comments.                                                                                     |

Reviewer 2: What do you mean by "available"? Is this word necessary?

Author: I think it does not add much. Changed to "The number of new comments."

Reviewer 2: Thanks.

### `RefreshHelperRenderless`

> A component that allows the application to refresh comment without the page reload.

Reviewer 2: ```suggestion
A component that allows the application to refresh a comment without page reload.
```

  >
    {({ pageNumber, comments }) => {
      // Replace with your code.
      console.log(
        `comments (pageNumber: ${pageNumber}, itemsOnPage: ${itemsOnPage}):`,
        comments
      )
      return null
    }}
  </CommentsPaginateRenderless>
)
```

The `CommentsPaginateRenderless` component takes the following props:

| Prop Name       | Description                                     |
| --------------- | ----------------------------------------------- |
| `itemsOnPage`   | The number of comments on a single page.        |
| `pageNumber`    | The number of the page to display.              |
| `readModelName` | The name of the comments read model.            |
| `resolverName`  | The name of the comments read model's resolver. |

This component passes all its props to its direct child.

The child component also receives the following additional props:

| Prop Name        | Description                                                            |
| ---------------- | ---------------------------------------------------------------------- |
| `comments`       | A list of comment data objects.                                        |

> | `paginationDone` | A `boolean` value that indicates if there is no more pages to display. |

Reviewer 2: ```suggestion
| `paginationDone` | A `boolean` value that indicates if there are no more pages to display. |
```

}
```

### `CommentsTreeRenderless`

A container used to render comments in a hierarchical view.

**Usage**

```js
import React from 'react'
import { CommentsTreeRenderless } from '@resolve-js/module-comments'

export const CommentsTree = (props) => (
  <CommentsTreeRenderless
    treeId="treeId"
    parentCommentId="parentCommentId"
    authorId="authorId"
    readModelName="customReadModelName" // default = 'readModelName'
    resolverName="customAllCommentsPaginate" // default = 'allCommentsPaginate'
  >
    {({ comments, createComment, updateComment, removeComment }) => {
      // Replace with your code.
      console.log('comments:', comments)
      return null
    }}
  </CommentsTreeRenderless>
)
```

The `CommentsTreeRenderless` component takes the following props:

| Prop Name         | Description                                                             |
| ----------------- | ----------------------------------------------------------------------- |
| `treeId`          | The unique identifier of a comment tree to display.                     |
| `parentCommentId` | The unique identifier of a comment for which to display child comments. |

> | `authorId`        | The unique identifier of an author whose comments to display.           |

Reviewer 2: ```suggestion
| `authorId`        | The unique identifier of an author whose comments are displayed.           |
```
